py-rse/py-scripting.Rmd

```diff
@@ -212,99 +212,222 @@ optional arguments:
+but more importantly ensures that
+the help message provided in response to the `-h` option
+will be the same as the interactive help.
+
+Let's try this out in an interactive Python session.
+(Remember, do not type the `>>>` prompt:
+Python provides this for us.)
 
 ```shell
-$ cd ~/zipfs_law/bin
+$ python
+Python 3.8.1 | packaged by conda-forge | (default, Jan 29 2020, 14:55:04)
+[GCC 7.3.0] on linux
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import script_template
+>>> script_template.__doc__
 ```
 
-Before we create our scripts for Zipf's Law,
-let's think through the tasks our scripts will need to accomplish.
+```text
+'One-line description of what the script does.'
+```
 
-To start,
-here's a function that uses the regular expressions (`re`) library
-to identify all the words in a text file and a counter function
-from the `collections` library to tally up the occurrences of each word.
+```shell
+>>> help(script_template)
+```
+
+```text
+Help on module script_template:
+
+NAME
+    script_template - One-line description of what the script does.
+
+FUNCTIONS
+    main(args)
+        Run the program.
+
+FILE
+    /Users/amira/script_template.py
+```
+
+As this example shows,
+if we ask for help on the module,
+Python formats and displays all of the docstrings for everything in the file.
+
+## Counting Words {#py-rse-py-scripting-wordcount}
+
+Now that we have a template for command-line Python programs,
+we can use it to check Zipf's Law for our collection of classic English novels.
+We start by moving the template into the directory
+where we store various programs and shell scripts we have created
+so that we can refer to it later for other projects:
+
+FIXME: where do we introduce `~/bin`?
+
+```shell
+$ mv script_template.py ~/bin
+```
+
+Next,
+let's write a function that counts how often words appear in a file.
+Our function splits the text on [whitespace][whitespace] characters
+(which is the default behavior of the string object's `split` method),
+then strips leading and trailing punctuation.
+This isn't completely correct---if two words are joined by a long dash
+like "correct" and "if" in this sentence, for example,
+they will be treated as one word---so
+we will explore better options in the exercises.
+We also use the `Counter` class from the `collections` library
+to count how many times each word occurs.
+If we give `Counter` a list of words,
+the result is an object that contains
+the number of times each one appears in the list:
 
 ```python
-import re
+import string
 from collections import Counter
 
 
 def count_words(reader):
     """Count the occurrence of each word in a string."""
     text = reader.read()
-    findwords = re.compile(r"\w+", re.IGNORECASE)
-    word_list = re.findall(findwords, text)
+    chunks = text.split()
+    stripped = [word.strip(string.punctuation) for word in chunks]
+    word_list = [word for word in stripped if word]
```

Comment:
```suggestion
    word_list = [word.lower() for word in stripped if word]
```

Reply:
I like this change to using string operators instead of regular expressions (as suggested by @joelostblom in #71). The only thing I'd suggest is a minor tweak to make all the letters lower case before counting the words. It makes a noticeable difference to the final word count.

Reply:
:+1: